README.md
# Mapster
This is a Google Map Api utility tool which is basically an OO abstraction of Google Map API, Google Place Auto-Complete, Marker-Clusterer

## Why should I care about this?
Think of this as a utility belt tool like Batman's but instead of catching the baddies, it's for handling Google Map Api, Google Place, and Marker Clusterer.

This is created as an OO high level absctraction tool that enables one to easily create a custom interactive map. 

By separating the logic of handling the map object, and the list of items / states, it makes the code of creating a custom map to be more portable and easier to re-use on other projects.

## Installation
ensure that the following script is included in your html page in the following order:
 ``` html
 <script src="https://maps.googleapis.com/maps/api/js?sensor=false&libraries=places"></script> // Google map api with Google Place library enabled.
 <script src="/mapster.min.js"></script>
```

## How do I use this?

### To create map
``` javascript
Mapster.create(selector, options)
```

This create a MAP var that initiate the creation of the MAPSTER object. The map is hooked on the DOM element with ID 'map-canvas' and options that can be configured. 

Example:
``` javascript
var MAP = Mapster.create('map-canvas', {
  center: {
    lat: -37.818667,
    lng: 144.971466
  },
  // further optional configuration setting for the map
  zoom: 10,
  disableDefaultUI: false,
  scrollwheel: true,
  draggable: true,
  mapTypeId: google.maps.MapTypeId.ROADMAP,
  zoomControlOptions: {
    position: google.maps.ControlPosition.LEFT_BOTTOM,
    style: google.maps.ZoomControlStyle.DEFAULT
  },
  panControlOptions: {
    position: google.maps.ControlPosition.LEFT_BOTTOM
  },
  cluster: {
    options: {
      styles: [{
        url: 'http://google-maps-utility-library-v3.googlecode.com/svn/trunk/markerclusterer/images/m2.png',
        height: 56,
        width: 55,
        textColor: '#F00',
        textSize: 18
      },{
        url: 'http://google-maps-utility-library-v3.googlecode.com/svn/trunk/markerclusterer/images/m1.png',
        height: 56,
        width: 55
      }]
    }
  },
  geocoder: true //activate geocoding function
});
```

### To hook Google Place AutoComplete
This will hook the Google Place Auto-Complete on the DOM input element with ID 'txtPlaces'. The snippet code below also passing in the options to attach events, and also limit the Google Autocomplete search.

``` javascript
  MAP.setPlaces('txtPlaces', {
    events: [{
      name: 'place_changed',
      callback: function(e, places){
        var place = places.getPlace();

        MAP.panTo({
          lat: place.geometry.location.lat(),
          lng: place.geometry.location.lng()
        });
      }
    }],
    places: {
      types: ['(regions)'],
      componentRestrictions: {country: "au"}
    }
  });

```

### To use HTML5 and get user's geolocation
This will utilize HTML5 navigator function to get the user's geolocation if the feature is enabled, and then pan the map to the user's location (or to the default location set in map's options if the feature is disabled)
```javascript
 MAP.getCurrentPosition( function(position) {
  MAP.panTo({
    lat: position.coords.latitude,
    lng: position.coords.longitude
  });
});
```
### Add marker on the map
This will add marker onto the map. We can pass in properties to the markers as object. The properties include event, custom styles, marker's properties, or additional properties we'd like to add onto the marker.
```javascript
 MAP.addMarker({
   lat: -37.818667,
   lng: 144.971466,
   animation: google.maps.Animation.DROP,
   content: 'Marker content',
});
```
### Get a list of marker 
The ```findBy``` function takes a callback that allows us to specify a condition. This returned callback will be passed to the List.js logic to filter the markers that satisfy the callback condition.
```javascript
var myGirlMarkers = MAP.findBy(function(marker) { 
    return marker.content === 'I like my girl'
});
```

### EXAMPLE
Refer to the example folder, or the [live demo link](http://jayzz55.github.io/mapster)

## List of Mapster's functions
* ``` zoom()``` = get current map's zoom level.
*  ``` zoom(num)``` = set current map's zoom level to specified num.
*  ``` panTo(coord)``` = pan the map to the specified coord position ```{lat: xx, lng: xx}```.
*  ```geocode(opts)``` = if ```geocoder``` = is enabled in map's options, we can pass in an address to the marker's ```address``` property to geocode the location through google map.
*  ```getCurrentPosition(callback)``` = Get user's current position. this function takes a callback that will be evaluated if the feature is enabled on the user's browser.
*  ```setPlaces(selector, opts)``` = Hook google place to the DOM element with id selector. by passing the callback we can specify the event we want to trigger, eg: on ```place_changed``` event.
*  ```addMarker(opts)``` = add maker onto the map, and passing in options.
*  ```findBy(callback)``` = find the markers that satisfy the condition in the evaluated callback function.
*  ```removeBy(callback)``` = remove the markers that satisfy the condition in the evaluated callback funciton.
*  ```removeAll()``` = remove all marker on the map.
*  ```visibleMarkersCount(selector)``` = count the visible markers on the map and render it on the DOM element with ID selector.

### PACKAGE
```
uglifyjs -c -m -o dist/mapster.min.js src/markerClustererPlus.js src/list.js src/mapster.js
```

## Contributors:
* [David East](https://github.com/davideast) - The main inspiration and the original mind behind mapster.
* [Andrew Buntine](https://github.com/buntine) - Refactoring and making mapster more awesome!

## Credits:
* Big thanks to David East, the library is built upon from one of David's course - [Tutsplus - Custom interactive maps with goole maps api](https://code.tutsplus.com/courses/custom-interactive-maps-with-the-google-maps-api)


CHANGELOG.md
# 0.1.2 / 2016-02-18

* Allow the options to pass in an MarkerClusterer options on Mapster init. Defaults to {ignoreHidden:true}

# 0.1.1 / 2015-08-18

* Fix bugs in the _remove, removeBy, and removeAll, and add feature to allow passing options into google Autocomplete

# 0.1.0 / 2015-08-14

* Released first polished mapster out to the world in a properly packaged library :)


src/list.js
(function (window) {
  var List = (function() {
    function List() {
      this.items = [];
    }

    List.prototype = {
      add: function(item) {
        this.items.push(item);
      },
      remove: function(item) {
        var indexOf = this.items.indexOf(item);
        if (indexOf !== -1) {
          this.items.splice(indexOf, 1);
        }
      },
      find: function(callback, action) {
        var items = this.items,
            length = items.length,
            matches = [],
            i = 0;

        for (; i < length; i++) {
          if (callback(items[i],i)) {
            matches.push(items[i]);
          }
        }

        if (action) {
          action.call(this, matches);
        }

        return matches;
      }
    };

    return List;
  } ());

  List.create = function(params) {
    return new List(params);
  };

  window.List = List;
} (window));


src/mapster.js
(function(window, google) {
  
  var Mapster = (function() {

    function Mapster(selector, opts) {
      var element = document.getElementById(selector);

      if (!opts.zoom) {
        opts.zoom = 16;
      }

      if (!opts.mapTypeId) {
        opts.mapTypeId = google.maps.MapTypeId.ROADMAP;
      }

      this.gMap = new google.maps.Map(element, opts);
      this.infoWindows = List.create();
      this.markers = List.create();

      if (opts.cluster) {
        var clusterOpts = opts.cluster.options ? opts.cluster.options : { ignoreHidden: true };
        this.markerClusterer = new MarkerClusterer(this.gMap, [], clusterOpts );
      }

      if (opts.geocoder) {
        this.geocoder = new google.maps.Geocoder();
      }
    }

    Mapster.prototype = {

      zoom: function(level) {
        if (level) {
          this.gMap.setZoom(level);
        }

        return this.gMap.getZoom();
      },

      panTo: function(coords) {
        this.gMap.panTo(coords);
      },

      _on: function(opts) {
        var self = this;
        google.maps.event.addListener(opts.obj, opts.event, function(e){
          opts.callback.call(self, e, opts.obj);
        });
      },

      geocode: function(opts) {
        this.geocoder.geocode({
          address: opts.address
        }, function(results, status) {
          if (status === google.maps.GeocoderStatus.OK) {
            opts.success.call(this, results, status);
          } else {
            opts.error.call(this, status);
          }
        });
      },

      getCurrentPosition: function(callback) {
        if (navigator.geolocation) {
          navigator.geolocation.getCurrentPosition(callback);
        }
      },

     setPlaces: function(selector, opts) {
        var element = document.getElementById(selector),
            places = new google.maps.places.Autocomplete(element, opts.places);
        
        this._attachEvents(places, opts.events);
      },

      addMarker: function(opts){
        var marker,
            infoWindow;

        opts.position = {
          lat: opts.lat,
          lng: opts.lng
        }

        marker = this._createMarker(opts);

        if (this.markerClusterer) {
          this.markerClusterer.addMarker(marker);  
        }
        
        this.markers.add(marker);

        if (opts.events) {
          this._attachEvents(marker, opts.events);
        }

        if (opts.content) {
          this._attachInfoWindow(opts, marker);
         }
        
        return marker;
      },

      _attachInfoWindow: function(opts, marker) {
        var infoWindow = new google.maps.InfoWindow({
          content: opts.content
        });
        
        this.infoWindows.add(infoWindow);
        
        this._on({
          obj: marker,
          event: 'click',
          callback: function() {
            this.infoWindows.items.forEach(function(infoWindow) {
              infoWindow.close();
            });

            infoWindow.open(this.gMap, marker);
          }
        })
      },

      _attachEvents: function(obj, events) {
        events.forEach(function(event) {
          this._on({
            obj: obj,
            event: event.name,
            callback: event.callback
          });
        }.bind(this));
      },

      findBy: function(callback) {
        return this.markers.find(callback);
      },

      removeBy: function(callback) {
        var set = this.markers.find(callback);
        this._remove(set);
      },

      removeAll: function() {
        this._remove(this.markers.items);
      },

      _remove: function(markers) {
        var mapster = this,
            set = this._clone(markers);

        set.forEach(function(marker) {
          if (mapster.markerClusterer) {
            mapster.markerClusterer.removeMarker(marker);
          } else {
            marker.setMap(null);              
          }

          mapster.markers.remove(marker);
        });

        return mapster.markers;
      },

      _clone: function(o) {
        var copy = Object.create(Object.getPrototypeOf(o));
        var propNames = Object.getOwnPropertyNames(o);

        propNames.forEach(function(name) {
          var desc = Object.getOwnPropertyDescriptor(o, name);
          Object.defineProperty(copy, name, desc);
        });

        return copy;
      },

      _createMarker: function(opts) {
        opts.map = this.gMap;
        return new google.maps.Marker(opts);
      },

      visibleMarkersCount: function(selector) {
        var visibleMarkers = this.findBy(function(marker) {
          return marker.getVisible();
        });

        document.getElementById(selector).innerHTML = visibleMarkers.length;
      }
    };

    return Mapster;

  }());
  
  Mapster.create = function(element, opts) {
    return new Mapster(element, opts);
  };
  
  window.Mapster = Mapster;
  
}(window, google));


example/gmap.js
var MAP,
    venueTypeChecked,
    venueDogFriendlyChecked;

$(document).ready(function() {

  // Initialize & Create the google map on DOM element ID 'map-canvas'.
  MAP = Mapster.create('map-canvas', {
    center: {
      lat: -37.818667,
      lng: 144.971466
    },
    cluster: true
  });

  // Hook the Google Place Auto-Complete utility on input with ID 'txtPlaces'.
  MAP.setPlaces('txtPlaces', {
    events: [{
      name: 'place_changed',
      callback: function(e, places){
        var place = places.getPlace();

        MAP.panTo({
          lat: place.geometry.location.G,
          lng: place.geometry.location.K
        });
      }
    }],
    places: {
      types: ['(regions)'],
      componentRestrictions: {country: "au"}
    }
  });

  // Using HTML5 navigator to get geolocation if the feature is available and then pan the map to the location.
  MAP.getCurrentPosition( function(position) {
    MAP.panTo({
      lat: position.coords.latitude,
      lng: position.coords.longitude
    });
  });

  seedData();

  MAP.visibleMarkersCount('total-result');

})

// seed data
function seedData() {
  for (var i = 0; i < 40; i++) {
    MAP.addMarker({
      lat: -37.818667 + Math.random()/50,
      lng: 144.971466 + Math.random()/50,
      animation: google.maps.Animation.DROP,
      content: 'I like my girl',
      venue_type: 'romantic', // Custom attributes.
      venue_dog_friendly: 'yes'
    });
    
    MAP.addMarker({
      lat: -37.818667 + Math.random()/50,
      lng: 144.971466 + Math.random()/50,
      animation: google.maps.Animation.DROP,
      content: 'You cant get any more awesome than this',
      venue_type: 'awesome',
      venue_dog_friendly: 'yes'
    }); 
  } 

  for (var i = 0; i < 20; i++) {
    MAP.addMarker({
      lat: -37.818667 + Math.random()/50,
      lng: 144.971466 + Math.random()/50,
      animation: google.maps.Animation.DROP,
      content: 'I like my girl',
      venue_type: 'romantic',
      venue_dog_friendly: 'no'
    });
    
    MAP.addMarker({
      lat: -37.818667 + Math.random()/50,
      lng: 144.971466 + Math.random()/50,
      animation: google.maps.Animation.DROP,
      content: 'You cant get any more awesome than this',
      venue_type: 'awesome',
      venue_dog_friendly: 'no'
    }); 
  }

  for (var i = 0; i < 10; i++) {
    MAP.addMarker({
      lat: -37.818667 + Math.random()/50,
      lng: 144.971466 + Math.random()/50,
      animation: google.maps.Animation.DROP,
      content: 'This place sucks',
      venue_type: 'boring',
      venue_dog_friendly: 'no'
    });
  }
};

// evaluate the OR condition across a name category
function evaluateCondition(base, collection) {
  var result = false;
  $.each(collection, function() {
    result = result || (base === $(this).val());
  });
  return result;
}

// add on 'Change' event handler on the input checkbox
$('input[type="checkbox"]').on('change', function() {
  venueTypeChecked = $('input[name="venue_type"]:checked');
  venueDogFriendlyChecked = $('input[name="venue_dog_friendly"]:checked');
       
  MAP.findBy(function(marker) {
    var totalCond = evaluateCondition(marker.venue_type, venueTypeChecked) &&
                    evaluateCondition(marker.venue_dog_friendly, venueDogFriendlyChecked);
        
    marker.setVisible(totalCond);
  });

  MAP.markerClusterer.repaint();

  MAP.visibleMarkersCount('total-result');

});
